File: crates/suspect-codegen/src/ruby_sdk/validation.rb
# frozen_string_literal: true

module __NAMESPACE__
  # @api private
  module Internal
    module_function

    def location(source) = source.fetch('document') + '#' + source.fetch('pointer')
    def child_path(path, key) = path + '/' + key.gsub('~', '~0').gsub('/', '~1')

    def metadata_integer(value)
      return value if value.instance_of?(Integer)
      raise EvaluationFailure, 'invalid integer metadata' unless value.instance_of?(JsonNumber)
      value.to_i(max_digits: 16)
    end

    # Numbers in literal enum/const operands stay JsonNumber. Only instruction
    # indices/cardinalities of the compiled metadata become Ruby Integers.
    def load_program(text)
      program = Json.parse(text, max_bytes: 16 * 1024 * 1024, max_depth: 128, max_work: 128 * 1024 * 1024)
      ProgramGuard.new(program).check
    rescue JsonError => error
      raise EvaluationFailure.new('invalid exact JSON program metadata', source: error.source), cause: error
    end

    def json_kind(value)
      return 'null' if value.nil?
      return 'boolean' if value.equal?(true) || value.equal?(false)
      return 'string' if value.instance_of?(String)
      return 'number' if value.instance_of?(Integer) || value.instance_of?(JsonNumber)
      return 'array' if value.instance_of?(Array)
      return 'object' if value.instance_of?(Hash)
      'invalid'
    end

    # One validation budget survives every logical trial and codec union dispatch.
    # The generator checked every opcode; an unknown opcode remains a runtime
    # evaluation failure, including when its ordinary instance kind is absent.
    class ValidationSession
      attr_reader :program

      def initialize(program = PROGRAM)
        # The shipped immutable program was checked once at package load. An
        # explicitly supplied program is independently checked without mutating
        # or freezing the caller's containers.
        program = Internal.load_program(Json.dump(program, max_bytes: 16 * 1024 * 1024, max_work: 128 * 1024 * 1024)) unless program.equal?(PROGRAM)
        @v2 = program['version'] != V1_VERSION
        @v3 = program['version'] == V3_VERSION
        @program, @nodes, @limits = program, program.fetch('nodes'), program.fetch('limits')
        @roots = program.fetch('roots').to_h { |root| [root.fetch('target'), true] }
        @steps, @equalities, @numeric = @limits.fetch('maxEvaluationSteps'), @limits.fetch('maxEqualitySteps'), @limits.fetch('maxEvaluationSteps')
        @findings, @active, @depth = [], {}, 0
        @numbers = {}
        @resource_context = @v3 ? program.fetch('resourceContext') : nil
        @resource_stack, @entered_resources, @contexts, @resource_identity = [], {}, {}, 0
      rescue JsonError => error
        raise EvaluationFailure.new('invalid exact JSON program metadata', source: error.source), cause: error
      end

      def failure(source, path, message)
        raise EvaluationFailure.new(message, source: Internal.location(source), instance_path: path)
      end

      def spend(source, path, amount = 1)
        @steps -= amount
        failure(source, path, 'schema evaluation work limit exceeded') if @steps.negative?
      end

      def mismatch(source, path, message)
        if @limits.fetch('maxErrors').zero? || @findings.length < @limits.fetch('maxErrors')
          @findings << ValidationError.new(message, source: Internal.location(source), instance_path: path)
        end
        false
      end

      def check(index, value)
        selected!(index)
        @findings = []
        valid = evaluate(index, value, '')
        raise(@findings.first || ValidationError.new('source schema rejected the value', source: Internal.location(@nodes.fetch(index).fetch('source')))) unless valid
        nil
      end

      def matches?(index, value, path = '')
        selected!(index)
        trial(index, value, path)
      end

      def selected!(index)
        raise EvaluationFailure, 'validation root was not selected' unless @roots.key?(index)
      end

      def number(value, source, path)
        token = value.instance_of?(String) ? value : Internal.number_token(value)
        failure(source, path, 'numeric operand byte limit exceeded') if token.bytesize > @limits.fetch('maxNumberBytes')
        return @numbers[token] if @v2 && @numbers.key?(token)
        @numeric -= token.bytesize
        failure(source, path, 'exact numeric work limit exceeded') if @numeric.negative?
        result = Exact.new(token)
        @numbers[token.dup.freeze] = result if @v2
        result
      rescue JsonError
        failure(source, path, 'invalid or oversized exact numeric operand')
      end

      def evaluate(index, value, path)
        return eval_scoped(index, value, path).first if @v2
        node = @nodes.fetch(index)
        source = node.fetch('source')
        spend(source, path)
        failure(source, path, 'schema evaluation depth limit exceeded') if @depth >= @limits.fetch('maxDepth')
        identity = [index, value.object_id]
        failure(source, path, 'nonproductive recursive schema evaluation') if @active[identity]
        @active[identity] = true
        @depth += 1
        begin
          valid = true
          node.fetch('checks').each do |check|
            # Deliberately evaluate even after a mismatch: later incomplete
            # evaluations cannot turn into completed invalidity.
            matched = instruction(check, value, path)
            valid = matched && valid
          end
          valid
        ensure
          @depth -= 1
          @active.delete(identity)
        end
      end

      def trial(index, value, path)
        findings, @findings = @findings, []
        begin
          evaluate(index, value, path)
        ensure
          @findings = findings
        end
      end

      def instruction(check, value, path)
        source, op = check.fetch('source'), check.fetch('op')
        spend(source, path)
        case op
        when 'ref' then evaluate(check.fetch('target'), value, path)
        when 'allOf', 'anyOf', 'oneOf'
          count = 0
          check.fetch('targets').each do |target|
            spend(source, path)
            count += 1 if op == 'allOf' ? evaluate(target, value, path) : trial(target, value, path)
          end
          valid = op == 'allOf' ? count == check['targets'].length : op == 'anyOf' ? count.positive? : count == 1
          valid || mismatch(source, path, "#{op} branch count rejected the value")
        when 'not'
          !trial(check.fetch('target'), value, path) || mismatch(source, path, 'negated schema accepted the value')
        when 'properties'
          valid = true
          if value.instance_of?(Hash)
            check.fetch('properties').each do |property|
              spend(source, path)
              name = property.fetch('name')
              next unless value.key?(name)
              matched = evaluate(property.fetch('target'), value[name], Internal.child_path(path, name))
              valid = matched && valid
            end
          end
          valid
        when 'additionalProperties'
          valid = true
          if value.instance_of?(Hash)
            value.each do |name, child|
              spend(source, path)
              next if check.fetch('declared').include?(name)
              matched = evaluate(check.fetch('target'), child, Internal.child_path(path, name))
              valid = matched && valid
            end
          end
          valid
        when 'items', 'prefixItems'
          valid = true
          if value.instance_of?(Array)
            start = op == 'items' ? check.fetch('start') : 0
            finish = op == 'items' ? value.length : [value.length, check.fetch('targets').length].min
            (start...finish).each do |i|
              spend(source, path)
              target = op == 'items' ? check.fetch('target') : check.fetch('targets').fetch(i)
              matched = evaluate(target, value[i], Internal.child_path(path, i.to_s))
              valid = matched && valid
            end
          end
          valid
        else scalar(check, value, path)
        end
      end

      def equal?(left, right, source, path)
        pending = [[left, right, 0]]
        until pending.empty?
          a, b, depth = pending.pop
          @equalities -= 1
          failure(source, path, 'structural equality work/depth limit exceeded') if @equalities.negative? || depth > @limits.fetch('maxDepth')
          kind = Internal.json_kind(a)
          return false unless kind == Internal.json_kind(b)
          case kind
          when 'number' then return false unless number(a, source, path).compare(number(b, source, path)).zero?
          when 'array'
            return false unless a.length == b.length
            failure(source, path, 'structural equality work limit exceeded') if pending.length + a.length > @equalities
            a.each_with_index { |child, i| pending << [child, b[i], depth + 1] }
          when 'object'
            return false unless a.length == b.length && a.keys.all? { |key| b.key?(key) }
            failure(source, path, 'structural equality work limit exceeded') if pending.length + a.length > @equalities
            a.each { |key, child| pending << [child, b[key], depth + 1] }
          when 'invalid' then failure(source, path, 'non-JSON value reached structural equality')
          else return false unless a == b
          end
        end
        true
      end

      def pattern?(program, text, source, path)
        raise EvaluationFailure, 'unsupported pattern version' unless program['version'] == 'suspect.pattern.experimental.v1'
        spend(source, path, text.length)
        codepoints, states = text.codepoints, program.fetch('states')
        (0..codepoints.length).each do |offset|
          current = [program.fetch('start')]
          (offset..codepoints.length).each do |position|
            pending, seen, consuming = current.dup, {}, []
            until pending.empty?
              spend(source, path)
              index = pending.pop
              next if seen[index]
              seen[index] = true
              state = states.fetch(index)
              case state.fetch('op')
              when 'match' then return true
              when 'split' then pending << state.fetch('first') << state.fetch('second')
              when 'jump' then pending << state.fetch('target')
              when 'start' then pending << state.fetch('target') if position.zero?
              when 'end' then pending << state.fetch('target') if position == codepoints.length
              when 'char' then consuming << index
              else failure(source, path, 'unknown portable pattern opcode')
              end
            end
            break if position == codepoints.length
            current = []
            consuming.each do |index|
              state = states.fetch(index)
              state.fetch('ranges').each do |low, high|
                spend(source, path)
                if codepoints[position].between?(low, high)
                  current << state.fetch('target')
                  break
                end
              end
            end
            break if current.empty?
          end
        end
        false
      end

      def scalar(check, value, path)
        source, op, kind = check.fetch('source'), check.fetch('op'), Internal.json_kind(value)
        valid = true
        case op
        when 'always' then valid = check.fetch('value').equal?(true)
        when 'type'
          types = check.fetch('types')
          valid = types.include?(kind) || kind == 'number' && types.include?('integer') && number(value, source, path).integral?
        when 'required'
          if kind == 'object'
            check.fetch('names').each do |name|
              spend(source, path)
              valid = mismatch(source, Internal.child_path(path, name), 'required property is absent') unless value.key?(name)
            end
          end
        when 'bound'
          if kind == 'number'
            order = number(value, source, path).compare(number(check.fetch('value'), source, path))
            valid = check.fetch('maximum') ? order.negative? : order.positive?
            valid ||= order.zero? && !check.fetch('exclusive')
          end
        when 'multipleOf'
          if kind == 'number'
            valid = number(value, source, path).multiple_of?(number(check.fetch('value'), source, path)) do |amount|
              @numeric -= amount
              failure(source, path, 'exact divisibility work limit exceeded') if @numeric.negative?
            end
          end
        when 'count'
          if kind == check.fetch('target')
            order = Exact.new(value.length.to_s).compare(number(check.fetch('value'), source, path))
            valid = check.fetch('maximum') ? order <= 0 : order >= 0
          end
        when 'const' then valid = equal?(value, check.fetch('value'), source, path)
        when 'enum'
          valid = false
          check.fetch('values').each do |literal|
            spend(source, path)
            if equal?(value, literal, source, path)
              valid = true
              break
            end
          end
        when 'uniqueItems'
          if kind == 'array'
            value.each_with_index do |child, i|
              spend(source, path)
              i.times do |j|
                spend(source, path)
                if equal?(child, value[j], source, path)
                  valid = false
                  break
                end
              end
              break unless valid
            end
          end
        when 'pattern' then valid = pattern?(check.fetch('program'), value, source, path) if kind == 'string'
        else failure(source, path, 'unknown portable validation opcode')
        end
        valid || mismatch(source, path, "#{op} assertion rejected the value")
      end
    end
  end
end

File: crates/suspect-codegen/src/typescript/validation.rs
//! Experimental TypeScript validation artifacts from a compiled owned program.
//!
//! This boundary never discovers or reinterprets OpenAPI keywords. It emits
//! only the proved static instructions supplied by `OwnedCompiler`.

use std::collections::{BTreeMap, BTreeSet};
use std::fmt;

use serde_json::{Value, json};
use suspect_schema::{OwnedProgram, ProgramInstruction, ProgramSource};

use crate::OutFile;

/// A portable program cannot be represented faithfully by this runtime.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationEmissionError {
    /// Original source identity, when one instruction or schema is responsible.
    pub source: Option<ProgramSource>,
    /// Located reason emission was refused.
    pub message: String,
}

impl fmt::Display for ValidationEmissionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for ValidationEmissionError {}

fn error(source: Option<&ProgramSource>, message: impl Into<String>) -> ValidationEmissionError {
    ValidationEmissionError {
        source: source.cloned(),
        message: message.into(),
    }
}

/// Emits a standalone exact JSON runtime, validator, compiled program and API
/// documentation. No model conversion or HTTP client is supplied by this seam.
///
/// The program must come from successful `OwnedCompiler` compilation. Public
/// snapshot fields are checked for supported version/profile, source identity,
/// safe metadata integers, valid target indices and numeric operands before
/// any artifact is returned. Enum/const values are parsed from lossless JSON
/// text in the generated module, never evaluated as numeric object literals.
pub fn emit(program: &OwnedProgram) -> Result<Vec<OutFile>, ValidationEmissionError> {
    check_program(program)?;
    let mut nodes = Vec::new();
    for node in &program.nodes {
        let mut checks = Vec::new();
        for check in &node.checks {
            let mut encoded = serde_json::to_value(check).expect("portable instruction serializes");
            let extra = match &check.instruction {
                ProgramInstruction::Enum { values } => {
                    encoded.as_object_mut().unwrap().remove("values");
                    Some(format!(
                        "\"values\":[{}]",
                        values.iter().map(operand).collect::<Vec<_>>().join(",")
                    ))
                }
                ProgramInstruction::Const { value } => {
                    encoded.as_object_mut().unwrap().remove("value");
                    Some(format!("\"value\":{}", operand(value)))
                }
                _ => None,
            };
            let mut text = encoded.to_string();
            if let Some(extra) = extra {
                text.pop();
                text.push(',');
                text.push_str(&extra);
                text.push('}');
            }
            checks.push(text);
        }
        nodes.push(format!(
            "{{\"source\":{},\"checks\":[{}]}}",
            serde_json::to_string(&node.source).unwrap(),
            checks.join(",")
        ));
    }
    let declarations = nodes
        .iter()
        .enumerate()
        .map(|(index, node)| {
            format!("const node{index}: NonNullable<ValidationProgram['nodes'][number]> = {node};")
        })
        .collect::<Vec<_>>()
        .join("\n");
    let all_nodes = (0..nodes.len())
        .map(|index| format!("node{index}"))
        .collect::<Vec<_>>()
        .join(",");
    let resource_declaration = program.resource_context.as_ref().map_or_else(
        String::new,
        |context| {
            format!(
                "\nconst resourceContext: NonNullable<ValidationProgram['resourceContext']> = {};",
                serde_json::to_string(context).unwrap()
            )
        },
    );
    let resource_field = if program.resource_context.is_some() {
        ",resourceContext"
    } else {
        ""
    };
    let mut root_validators = String::new();
    for (index, root) in program.roots.iter().enumerate() {
        // V3 bindings may target a node outside the ordinary static closure.
        // Retain its complete checked resource registry; unentered candidates
        // remain inert at execution. Frozen v1/v2 sparse root slices are unchanged.
        let reachable = if program.resource_context.is_some() {
            (0..nodes.len()).collect()
        } else {
            reachable_nodes(program, root.target)
        };
        let table = reachable
            .iter()
            .map(|target| format!("{target}:node{target}"))
            .collect::<Vec<_>>()
            .join(",");
        let root_notice = if program.resource_context.is_some() {
            "Validate this selected root with the checked resource registry and runtime-selected bindings.".to_owned()
        } else {
            format!("Validate only the finite node closure reachable from selected root {index}.")
        };
        root_validators.push_str(&format!(
            "/** {root_notice} */\nexport const validateRoot{index} = /* @__PURE__ */ createLazyValidator(() => createValidator({{version:{},profile:{},roots:[{}],nodes:Object.assign([],{{{table}}}) as ValidationProgram['nodes'],limits:{}{resource_field}}}));\n",
            quoted(program.version), quoted(program.profile), serde_json::to_string(root).unwrap(), serde_json::to_string(&program.limits).unwrap()
        ));
    }
    let module = format!(
        "// Generated from a proved compiled validation program.\nimport {{ parseJson }} from './json.js';\nimport {{ createLazyValidator, createValidator, type ValidationProgram }} from './validation.js';\n\n{declarations}{resource_declaration}\n\nconst program: ValidationProgram = {{\nversion:{},\nprofile:{},\nroots:{},\nnodes:[{all_nodes}],\nlimits:{}{resource_field}\n}};\n\n/** Selected original OpenAPI schema identities. */\nexport const validationRoots = /* @__PURE__ */ (() => Object.freeze(program.roots.map(root => Object.freeze({{...root.source}}))))();\n/** Validate exact wire values; does not decode language models or send HTTP. */\nexport const validate = /* @__PURE__ */ createLazyValidator(() => createValidator(program));\n{root_validators}",
        quoted(program.version),
        quoted(program.profile),
        serde_json::to_string(&program.roots).unwrap(),
        serde_json::to_string(&program.limits).unwrap(),
    );
    Ok(vec![
        super::json::runtime(),
        OutFile {
            path: "typescript/validation.ts".into(),
            content: include_str!("validation.ts").into(),
        },
        OutFile {
            path: "typescript/pattern.ts".into(),
            content: include_str!("pattern.ts").into(),
        },
        OutFile {
            path: "typescript/validation-resources.ts".into(),
            content: include_str!("validation-resources.ts").into(),
        },
        OutFile {
            path: "typescript/uri.ts".into(),
            content: include_str!("uri.ts").into(),
        },
        OutFile {
            path: "typescript/validation-program.ts".into(),
            content: module,
        },
        OutFile {
            path: "typescript/validation.md".into(),
            content: documentation(program),
        },
    ])
}

fn quoted(text: &str) -> String {
    serde_json::to_string(text)
        .unwrap()
        .replace('\u{2028}', "\\u2028")
        .replace('\u{2029}', "\\u2029")
}

fn operand(value: &Value) -> String {
    let text = value.to_string();
    // UTF-8 byte length safely bounds UTF-16 units, JSON nesting, node count,
    // and numeric token length. Representation parsing must not pre-apply the
    // separate numeric/equality schema limits to an unvisited literal operand.
    let cap = text.len();
    format!(
        "/* @__PURE__ */ parseJson({},{{maxLength:{cap},maxDepth:{cap},maxNodes:{cap},maxNumberLength:{cap}}})",
        quoted(&text)
    )
}

fn reachable_nodes(program: &OwnedProgram, root: usize) -> BTreeSet<usize> {
    let mut reached = BTreeSet::new();
    let mut pending = vec![root];
    while let Some(index) = pending.pop() {
        if !reached.insert(index) {
            continue;
        }
        for check in &program.nodes[index].checks {
            match &check.instruction {
                ProgramInstruction::Ref { target }
                | ProgramInstruction::Not { target }
                | ProgramInstruction::AdditionalProperties { target, .. }
                | ProgramInstruction::AdditionalPropertiesWithPatterns { target, .. }
                | ProgramInstruction::Contains { target, .. }
                | ProgramInstruction::PropertyNames { target }
                | ProgramInstruction::UnevaluatedProperties { target }
                | ProgramInstruction::UnevaluatedItems { target }
                | ProgramInstruction::Items { target, .. } => pending.push(*target),
                ProgramInstruction::Properties { properties }
                | ProgramInstruction::DependentSchemas {
                    dependencies: properties,
                } => pending.extend(properties.iter().map(|property| property.target)),
                ProgramInstruction::If {
                    condition,
                    then_target,
                    else_target,
                } => {
                    pending.push(*condition);
                    pending.extend(then_target);
                    pending.extend(else_target);
                }
                ProgramInstruction::PatternProperties { patterns } => {
                    pending.extend(patterns.iter().map(|(_, _, target)| *target))
                }
                ProgramInstruction::PrefixItems { targets }
                | ProgramInstruction::AllOf { targets }
                | ProgramInstruction::AnyOf { targets }
                | ProgramInstruction::OneOf { targets } => pending.extend(targets.iter().copied()),
                _ => {}
            }
        }
    }
    reached
}

/// The checked root closure used by a native codec, with graph-local indices.
/// Locations are retained by the containing model record, not interface equality.
/// Only instruction targets are remapped: enum/const instance data is opaque.
pub(super) fn interface(program: &OwnedProgram, root: usize) -> Value {
    let mut indices = BTreeMap::from([(root, 0)]);
    let mut pending = vec![root];
    let mut resource_indices = BTreeMap::new();
    let mut resources = Vec::new();
    let mut scopes = Vec::new();
    let mut nodes = Vec::new();
    let mut next = 0;
    while next < pending.len() {
        let index = pending[next];
        next += 1;
        if let Some(context) = &program.resource_context {
            let resource = context.node_scopes[index].0;
            scopes.push(*resource_indices.entry(resource).or_insert_with(|| {
                resources.push(resource);
                resources.len() - 1
            }));
            for (_, _, target) in &context.resources[resource].dynamic_anchors {
                indices.entry(*target).or_insert_with(|| {
                    pending.push(*target);
                    pending.len() - 1
                });
            }
        }
        let mut checks = Vec::new();
        for check in &program.nodes[index].checks {
            let mut instruction = check.instruction.clone();
            let mut target = |value: &mut usize| {
                let index = *indices.entry(*value).or_insert_with(|| {
                    pending.push(*value);
                    pending.len() - 1
                });
                *value = index;
            };
            match &mut instruction {
                ProgramInstruction::DynamicRef {
                    target: value,
                    initial_resource,
                    ..
                } => {
                    target(value);
                    *initial_resource =
                        *resource_indices
                            .entry(*initial_resource)
                            .or_insert_with(|| {
                                resources.push(*initial_resource);
                                resources.len() - 1
                            });
                }
                ProgramInstruction::Ref { target: value }
                | ProgramInstruction::Not { target: value }
                | ProgramInstruction::AdditionalProperties { target: value, .. }
                | ProgramInstruction::AdditionalPropertiesWithPatterns { target: value, .. }
                | ProgramInstruction::Items { target: value, .. }
                | ProgramInstruction::Contains { target: value, .. }
                | ProgramInstruction::PropertyNames { target: value }
                | ProgramInstruction::UnevaluatedProperties { target: value }
                | ProgramInstruction::UnevaluatedItems { target: value } => target(value),
                ProgramInstruction::Properties { properties }
                | ProgramInstruction::DependentSchemas {
                    dependencies: properties,
                } => {
                    for property in properties {
                        target(&mut property.target);
                    }
                }
                ProgramInstruction::If {
                    condition,
                    then_target,
                    else_target,
                } => {
                    target(condition);
                    for value in then_target.iter_mut().chain(else_target) {
                        target(value);
                    }
                }
                ProgramInstruction::PatternProperties { patterns } => {
                    for (_, _, value) in patterns {
                        target(value);
                    }
                }
                ProgramInstruction::PrefixItems { targets }
                | ProgramInstruction::AllOf { targets }
                | ProgramInstruction::AnyOf { targets }
                | ProgramInstruction::OneOf { targets } => {
                    for value in targets {
                        target(value);
                    }
                }
                _ => {}
            }
            checks.push(instruction);
        }
        nodes.push(json!({"checks":checks}));
    }
    let mut value = json!({"version":program.version,"profile":program.profile,"root":0,"nodes":nodes,"limits":program.limits});
    if let Some(context) = &program.resource_context {
        value["resourceContext"] = json!({"nodeScopes":scopes,"resources":resources.iter().map(|resource| {
            let resource = &context.resources[*resource];
            json!({"kind":resource.kind,"dynamicAnchors":resource.dynamic_anchors.iter().map(|(name,_,target)|json!({"name":name,"target":indices[target]})).collect::<Vec<_>>()})
        }).collect::<Vec<_>>()});
    }
    value
}

fn safe(
    value: usize,
    source: Option<&ProgramSource>,
    name: &str,
) -> Result<(), ValidationEmissionError> {
    if value as u128 > 9_007_199_254_740_991 {
        return Err(error(
            source,
            format!("{name} is not a JavaScript safe integer"),
        ));
    }
    Ok(())
}

fn check_program(program: &OwnedProgram) -> Result<(), ValidationEmissionError> {
    program.check().map_err(|error| ValidationEmissionError {
        source: error.source,
        message: error.message,
    })?;
    if !matches!(
        (program.version, program.profile),
        (OwnedProgram::V1_VERSION, OwnedProgram::V1_PROFILE)
            | (OwnedProgram::V2_VERSION, OwnedProgram::V2_PROFILE)
            | (OwnedProgram::V3_VERSION, OwnedProgram::V3_PROFILE)
    ) {
        return Err(error(
            None,
            "unsupported TypeScript validation version/profile",
        ));
    }
    for (name, value) in [
        ("maxDepth", program.limits.max_depth),
        ("maxErrors", program.limits.max_errors),
        ("maxNumberBytes", program.limits.max_number_bytes),
        ("maxEqualitySteps", program.limits.max_equality_steps),
        ("maxEvaluationSteps", program.limits.max_evaluation_steps),
        ("node count", program.nodes.len()),
    ] {
        safe(value, None, name)?;
    }
    for root in &program.roots {
        safe(root.target, Some(&root.source), "root target")?;
    }
    if let Some(context) = &program.resource_context {
        safe(context.resources.len(), None, "resource count")?;
        for (index, source, _) in &context.node_scopes {
            safe(*index, Some(source), "resource index")?;
        }
        for resource in &context.resources {
            for (_, source, target) in &resource.dynamic_anchors {
                safe(*target, Some(source), "dynamic anchor target")?;
            }
        }
    }
    for node in &program.nodes {
        for check in &node.checks {
            let at = Some(&check.source);
            match &check.instruction {
                ProgramInstruction::DynamicRef {
                    target,
                    initial_resource,
                    ..
                } => {
                    safe(*target, at, "dynamic target")?;
                    safe(*initial_resource, at, "initial resource")?;
                }
                ProgramInstruction::Ref { target }
                | ProgramInstruction::Not { target }
                | ProgramInstruction::AdditionalProperties { target, .. }
                | ProgramInstruction::AdditionalPropertiesWithPatterns { target, .. }
                | ProgramInstruction::Contains { target, .. }
                | ProgramInstruction::PropertyNames { target }
                | ProgramInstruction::UnevaluatedProperties { target }
                | ProgramInstruction::UnevaluatedItems { target } => {
                    safe(*target, at, "schema target")?
                }
                ProgramInstruction::Properties { properties }
                | ProgramInstruction::DependentSchemas {
                    dependencies: properties,
                } => {
                    for property in properties {
                        safe(property.target, at, "property target")?;
                    }
                }
                ProgramInstruction::If {
                    condition,
                    then_target,
                    else_target,
                } => {
                    safe(*condition, at, "condition target")?;
                    for target in then_target.iter().chain(else_target) {
                        safe(*target, at, "conditional target")?;
                    }
                }
                ProgramInstruction::PatternProperties { patterns } => {
                    for (_, _, target) in patterns {
                        safe(*target, at, "pattern property target")?;
                    }
                }
                ProgramInstruction::Items { target, start } => {
                    safe(*target, at, "items target")?;
                    safe(*start, at, "items start")?;
                }
                ProgramInstruction::PrefixItems { targets }
                | ProgramInstruction::AllOf { targets }
                | ProgramInstruction::AnyOf { targets }
                | ProgramInstruction::OneOf { targets } => {
                    for target in targets {
                        safe(*target, at, "branch target")?;
                    }
                }
                _ => {}
            }
        }
    }
    Ok(())
}

fn documentation(program: &OwnedProgram) -> String {
    let mut text = String::from(
        "# Compiled TypeScript validation\n\nThis experimental validator executes the compiled OAS 3.1 / JSON Schema 2020-12 static subset. It does not decode language models or provide an HTTP client.\n\n```ts\nimport { parseJson } from './json.js';\nimport { validate, validationRoots } from './validation-program.js';\nconst result = validate(validationRoots[0]!, parseJson(responseText));\nif (result.kind === 'invalid') console.error(result.findings);\nif (result.kind === 'evaluationFailure') console.error(result.finding);\n```\n\n`valid` means every compiled assertion passed. `invalid` contains capped schema mismatches. `evaluationFailure` means validity could not be determined; logical branches cannot suppress or invert it. Each finding carries the original document URI, schema pointer and RFC 6901 instance pointer. Only selected roots may be requested.\n\nNumbers remain exact `JsonNumber` values. No defaults, coercion, model conversion or unknown-field stripping occurs. Pass values from `parseJson`, or construct values satisfying `WireJsonValue`. Trusted generated modules and JavaScript builtins must remain intact; this API is not isolation from application mutation of imported code.\n\nLimits apply independently to each call. The compiled visit budget covers schema, keyword, branch and collection visits; equality and numeric operand limits remain separate. These are visit/representation limits, not a total byte-work or allocation bound.\n\nThe optional third argument is a trace observer receiving `(source, instancePath, valid)` after each completed schema application, including trials. It receives no incomplete application, must not mutate the instance, and does not receive a fresh budget. Use traces only after the overall result is valid. Observer exceptions produce evaluation failure.\n\nSelected sources:\n\n",
    );
    for root in &program.roots {
        text.push_str(&format!(
            "- <code>{}</code> at <code>{}</code>\n",
            inert_text(&root.source.document),
            inert_text(&root.source.pointer)
        ));
    }
    if program.version == OwnedProgram::V2_VERSION {
        text.push_str(SCOPED_DOCUMENTATION);
    } else if program.version == OwnedProgram::V3_VERSION {
        text.push_str(RESOURCE_DOCUMENTATION);
    }
    text
}

pub(super) const SCOPED_DOCUMENTATION: &str = "\n## Scoped v2 validation\n\nThe checked `suspect.validation.experimental.v2` / `oas31-jsonschema202012-static-applicators` program implements if/then/else, dependentRequired, dependentSchemas, contains with exact minContains/maxContains, patternProperties and pattern-aware additionalProperties, propertyNames, unevaluatedProperties and unevaluatedItems. Each child starts with fresh evaluated-property/item sets; only the documented successful sets propagate. Conditional branches and dependent schemas are not flattened into property lists. Numeric, equality, depth and shared work failures remain evaluation failures through every trial and logical branch. Collection visits, NFA transitions and annotation-set merge candidates spend the same per-call work allowance.\n\nDeclared native fields retain their types. Pattern-matched extras and heterogeneous prefixItems use a checked JsonValue carrier where static types cannot express the constraint. Values retain exact numbers, all permitted keys, absence and null. Decode validates the source schema; encode validates the current wire value again after mutation. Required properties, pattern overlaps, key-name constraints and unevaluated locations are enforced by the codec. Ordinary base closures keep the v1 program. Resource/dynamic schema programs require their separately verified executable profile.\n";

pub(super) const RESOURCE_DOCUMENTATION: &str = "\n## Resources and dynamic references\n\nThe checked `suspect.validation.experimental.v3` / `oas31-jsonschema202012-resources-dynamic` program retains the indexed physical resource boundaries, logical URI names and finite dynamic bindings. The runtime performs no source loading, URI lookup or acquisition. Entering a nested schema enters its indexed resource without evaluating an unselected resource root. The outermost actually entered matching resource wins; unentered candidates are inert. Pointer, empty-fragment and ordinary-anchor fallbacks remain static. Every return and trial restores resource scope. Context-sensitive cycle identities and per-call resource-entry/binding-scan work are exact; failures cannot be inverted or suppressed.\n\nV3 includes the scoped v2 applicators. A dynamically selected target starts with fresh evaluated-property/item sets and propagates only successful annotations. Dynamic positions use checked JsonValue carriers instead of assuming the initial target's native type. Known fields retain their native representations, exact numbers and absence/null distinctions. Decode validates the selected root; encode validates the current wire value again after mutation. Model-only rendering retains explicit codec obligations. Ordinary closures keep their established v1/v2 programs; the planner selects compile_v3 only for source-required resource semantics.\n\nHTTP relative servers use their effective physical retrieval document, independently of logical $self/$id names. Caller documentURL overrides stay explicit. Credential callbacks receive effectiveServerURL for relative OAuth/OIDC endpoint metadata; the SDK does not acquire those endpoints. Native Fetch rewrites of encoded dot segments are refused before sending; a capable explicit transport receives the exact URI spelling.\n";

fn inert_text(text: &str) -> String {
    text.chars()
        .map(|ch| {
            if ch.is_alphanumeric() || ch == ' ' {
                ch.to_string()
            } else {
                format!("&#{};", u32::from(ch))
            }
        })
        .collect()
}
